### PR TITLE
US114642 - feedback UI

### PR DESCRIPTION
**Pull request by jeroach**
initial setup to use components from FACE

What this looks like:
Annotation 2020-05-14 115515

### CODE REVIEW

File: components/right-panel/consistent-evaluation-feedback.js [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+				this._initializeController().then(() => this.requestUpdate());
+			}
+		}
+	}
+
+	async _initializeController() {
+		this._controller = new ConsistentEvaluationFeedbackController(this._href, this._token);
+
+		this._feedbackEntity = await this._controller.requestFeedback();
+		this._feedbackText = this._feedbackEntity.entities[0].entities[0].properties.text;
```
Comment (jeroach) [inline components/right-panel/consistent-evaluation-feedback.js:76]: This is mega ugly - but we have not been able to get the entity store to work

File: components/right-panel/consistent-evaluation-feedback.js [hunk 2]
```diff
@@ -0,0 +1,121 @@
+import 'd2l-activities/components/d2l-activity-editor/d2l-activity-text-editor.js';
```
Comment (jeroach) [inline components/right-panel/consistent-evaluation-feedback.js:1]: Mainly looking for thoughts/comments on this file

File: components/right-panel/consistent-evaluation-feedback.js [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+			timeOut.after(500),
+			() => this._saveFeedback(feedback, this._feedbackEntity)
+		);
+	}
+
+	render() {
+		return html`
+			<d2l-consistent-evaluation-right-panel-block title=${this.header}>
+				<div id="feedback-container">
+					<d2l-activity-text-editor
```
Comment (jeroach) [inline components/right-panel/consistent-evaluation-feedback.js:101]: FACE html editor

File: components/right-panel/consistent-evaluation-feedback.js [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+	static async getLocalizeResources(langs) {
+		return await loadLocalizationResources(langs);
+	}
+
+	constructor() {
+		super();
+
+		this._feedbackText = '';
+		this._debounceJobs = {};
+
```
Comment (dmackenz) [inline components/right-panel/consistent-evaluation-feedback.js:40]: may want to set `this._href = undefined` and `this._token = undefined`?

File: components/right-panel/consistent-evaluation-feedback.js [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+						.value="${this._feedbackText}"
+						.richtextEditorConfig="${this.richTextEditorConfig}"
+						@d2l-activity-text-editor-change="${this._saveOnFeedbackChange}"
+						ariaLabel="${this.localize('overallFeedback')}"
+						?disabled="${!this.canEditFeedback}">
+					</d2l-activity-text-editor>
+				</div>
+				
+				<div id="feedback-actions-container" ?hidden="${!this.href}">
+					<d2l-activity-attachments-editor
```
Comment (jeroach) [inline components/right-panel/consistent-evaluation-feedback.js:111]: FACE feedback actions

File: components/right-panel/consistent-evaluation-feedback.js [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+		this._feedbackText = this._feedbackEntity.entities[0].entities[0].properties.text;
+	}
+
+	_saveFeedback(feedback, entity) {
+		this._debounceJobs.value = Debouncer.debounce(
+			this._debounceJobs.value,
+			timeOut.after(2000),
+			() => this._controller.updateFeedbackText(feedback, entity)
+		);
+	}
```
Comment (dmackenz) [inline components/right-panel/consistent-evaluation-feedback.js:85]: Not really sure what the debouncer is for here? Do you just want to call a function and make it timeout after a certain amount of time?

File: components/right-panel/consistent-evaluation-right-panel.js [hunk 7]
```diff
@@ -92,15 +78,12 @@ export class ConsistentEvaluationRightPanel extends LocalizeMixin(LitElement) {
...
-					value="This is the value"
-					ariaLabel="aria label"
+					href=${this.feedbackHref}
 					.richtextEditorConfig=${this._richTextEditorConfig}
-					?disabled=${this.richTextEditorDisabled}
-					@d2l-request-provider=${this._onRequestProvider}
-					@html-editor-demo-change=${this._saveInstructionsOnChange}
-				></d2l-consistent-evaluation-html-editor>
+					token=${this.token}
+				></d2l-consistent-evaluation-feedback>
```
Comment (dmackenz) [inline components/right-panel/consistent-evaluation-right-panel.js:86]: much cleaner 👍

File: demo/consistent-evaluation-feedback.html [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+
+			<d2l-demo-snippet>
+				<d2l-consistent-evaluation-feedback
+					can-edit-feedback
+					header="Feedback"
+					href="https://a5fec3f6-6ff3-44e9-9323-5f94785b26d3.activities.api.dev.brightspace.com/activities/6606_2000_1/usages/6609/users/178/feedback"
+					.richtextEditorConfig="{}"
+					token="eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiIsImtpZCI6IjY4ZTE3ZTgyLTJjZjUtNDE1MC05NDQ4LTlkZmU5MmFiY2U2NCJ9.eyJpc3MiOiJodHRwczovL2FwaS5icmlnaHRzcGFjZS5jb20vYXV0aCIsImF1ZCI6Imh0dHBzOi8vYXBpLmJyaWdodHNwYWNlLmNvbS9hdXRoL3Rva2VuIiwiZXhwIjoxNTg5NDcyNjI2LCJuYmYiOjE1ODk0NjkwMjYsInN1YiI6IjMwMjIwIiwidGVuYW50aWQiOiJhNWZlYzNmNi02ZmYzLTQ0ZTktOTMyMy01Zjk0Nzg1YjI2ZDMiLCJhenAiOiJsbXM6YTVmZWMzZjYtNmZmMy00NGU5LTkzMjMtNWY5NDc4NWIyNmQzIiwic2NvcGUiOiIqOio6KiIsImp0aSI6ImY3NjA2OTYyLWRlMGQtNDBlMy05N2VlLTVmOTA2YmQ0YzhmYiJ9.WY92dx6sDsI2tQRShCF3rNYbFfXEBuP1r3jDukjxkRdJrZsrDMgnzSOWiKBpt1Bks1e_8OpFc4XujeI_ugRVf63aDWjd3QPl1_XA8HfyeDVFb8ZDbJenScg__AJ6Zc0yki_gTt0F5YuF-ELkbN4LeOAWoWJMumkWp3feRWwM_QrbIjrqsPdZZrSWa5KZ18nxtDGBfhKT797pE9x_DWDzNY0QR0uEo7tzvMuWov97Y-7ig6rKt-Gu8BC2MWNli72miwt3XlqF7Bsj5KqrlaCPDtgvQ71ikrVfq5UG5Vyv0ccHuIaxx2fD_jrbRowxUvvqaNpcmVnfdD7SPnkuDgIdvw"
```
Comment (dmackenz) [inline demo/consistent-evaluation-feedback.html:26]: watch out committing the hrefs and tokens in the future... It probably doesn't matter but something to note

File: components/right-panel/consistent-evaluation-feedback.js (hunk 1 above)
Comment (dmackenz) [reply components/right-panel/consistent-evaluation-feedback.js:76]: Could you use one of these functions instead? https://github.com/Brightspace/node-siren-parser#entitygetxbyystringregexp-key

File: components/right-panel/consistent-evaluation-feedback.js (hunk 6 above)
Comment (vix633) [reply components/right-panel/consistent-evaluation-feedback.js:85]: debounce here so the save route isnt called on every key press.

File: demo/consistent-evaluation-feedback.html (hunk 8 above)
Comment (vix633) [reply demo/consistent-evaluation-feedback.html:26]: These are long expired, site is ephemeral as well. No worries here. But yeah, you suck josh

Comment (jeroach) [reply demo/consistent-evaluation-feedback.html:26]: I do suck - sorry guys - my bad

File: components/right-panel/consistent-evaluation-right-panel.js (hunk 7 above)
Comment (jeroach) [reply components/right-panel/consistent-evaluation-right-panel.js:86]: Thanks!

File: components/right-panel/consistent-evaluation-feedback-presentational.js [hunk 9]
```diff
@@ -1,66 +1,66 @@
-import 'd2l-activities/components/d2l-activity-editor/d2l-activity-text-editor.js';
```
Comment (jeroach) [inline components/right-panel/consistent-evaluation-feedback-presentational.js]: Hmmm, might want to check your line ending/whitespace settings @zhangium

File: components/right-panel/consistent-evaluation-feedback-presentational.js [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+import { css, html, LitElement } from 'lit-element';
+import { loadLocalizationResources } from '../locale.js';
+import { LocalizeMixin } from '@brightspace-ui/core/mixins/localize-mixin.js';
+
+class ConsistentEvaluationFeedbackPresentational extends LocalizeMixin(LitElement) {
+	static get properties() {
+		return {
+			canEditFeedback: {
+				attribute: 'can-edit-feedback',
+				reflect: true,
```
Comment (AlexBedley) [inline components/right-panel/consistent-evaluation-feedback-presentational.js:13]: i dont think we need attributes, maybe try and change to using only properties (ie `.canEditFeedback=true`)

File: components/right-panel/consistent-evaluation-feedback-presentational.js [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+				type: String
+			},
+			href: { type: String },
+			richTextEditorConfig: { type: Object },
+			saveOnFeedbackChange: { type: Object },
+			token: { type: String }
+		};
+	}
+
+	static get styles() {
```
Comment (AlexBedley) [inline components/right-panel/consistent-evaluation-feedback-presentational.js:28]: nit: is this necessary?

File: components/right-panel/consistent-evaluation-feedback-presentational.js [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+	constructor() {
+		super();
+
+		this.canEditFeedback = false;
+	}
+
+	render() {
+		return html`
+			<d2l-consistent-evaluation-right-panel-block title="${this.localize('overallFeedback')}">
+				<div id="feedback-container">
```
Comment (AlexBedley) [inline components/right-panel/consistent-evaluation-feedback-presentational.js:45]: nit: don't have ids (or classes) that aren't used

File: components/right-panel/consistent-evaluation-feedback.js [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+import { ConsistentEvaluationFeedbackController } from '../controllers/ConsistentEvaluationFeedbackController.js';
+import { Debouncer } from '@polymer/polymer/lib/utils/debounce.js';
+import { timeOut } from '@polymer/polymer/lib/utils/async.js';
+
+class ConsistentEvaluationFeedback extends LitElement {
+	static get properties() {
+		return {
+			href: { type: String },
+			token: { type: String }
+			//_debounceJobs: { type: Object }
```
Comment (AlexBedley) [inline components/right-panel/consistent-evaluation-feedback.js:12]: nit: kill it with fire

File: components/right-panel/consistent-evaluation-right-panel.js [hunk 14]
```diff
@@ -92,15 +78,10 @@ export class ConsistentEvaluationRightPanel extends LocalizeMixin(LitElement) {
...
-					value="This is the value"
-					ariaLabel="aria label"
-					.richtextEditorConfig=${this._richTextEditorConfig}
-					?disabled=${this.richTextEditorDisabled}
-					@d2l-request-provider=${this._onRequestProvider}
-					@html-editor-demo-change=${this._saveInstructionsOnChange}
-				></d2l-consistent-evaluation-html-editor>
+				<d2l-consistent-evaluation-feedback
+					href=${this.feedbackHref}
+					token=${this.token}
```
Comment (AlexBedley) [inline components/right-panel/consistent-evaluation-right-panel.js:83]: need `.token`

File: components/right-panel/consistent-evaluation-feedback.js [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+
+	render() {
+		return html`
+			<d2l-consistent-evaluation-feedback-presentational
+				can-edit-feedback
+				feedback-text="${this._feedbackText}"
+				href="${this._href}"
+				.richtextEditorConfig="${this._richTextEditorConfig}"
+				@d2l-activity-text-editor-change="${this._saveOnFeedbackChange}"
+				token="${this._token}"
```
Comment (AlexBedley) [inline components/right-panel/consistent-evaluation-feedback.js:85]: `.token`

File: components/right-panel/consistent-evaluation-feedback-presentational.js [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+	}
+
+	static async getLocalizeResources(langs) {
+		return await loadLocalizationResources(langs);
+	}
+
+	constructor() {
+		super();
+
+		this.canEditFeedback = true;
```
Comment (jeroach) [inline components/right-panel/consistent-evaluation-feedback-presentational.js:31]: This should default to `false` as per: https://blog.d2l.dev/2020/05/11/LitElement-Best-Practices-Gotchas/#-do-default-boolean-properties-to-false

File: components/right-panel/consistent-evaluation-feedback-presentational.js [hunk 17]
```diff
@@ -0,0 +1,58 @@
...
+import './consistent-evaluation-right-panel-block';
+import { html, LitElement } from 'lit-element';
+import { loadLocalizationResources } from '../locale.js';
+import { LocalizeMixin } from '@brightspace-ui/core/mixins/localize-mixin.js';
+
+class ConsistentEvaluationFeedbackPresentational extends LocalizeMixin(LitElement) {
+	static get properties() {
+		return {
+			canEditFeedback: { type: Boolean },
+			feedbackText: {
```
Comment (AlexBedley) [inline components/right-panel/consistent-evaluation-feedback-presentational.js:12]: I think this still needs the `canEditFeedback` treatment - I don't think we need to reflect to attribute

File: components/right-panel/consistent-evaluation-feedback-presentational.js [hunk 18]
```diff
@@ -0,0 +1,58 @@
...
+	constructor() {
+		super();
+
+		this.canEditFeedback = true;
+	}
+
+	render() {
+		return html`
+			<d2l-consistent-evaluation-right-panel-block title="${this.localize('overallFeedback')}">
+				<div>
```
Comment (AlexBedley) [inline components/right-panel/consistent-evaluation-feedback-presentational.js:37]: I know I just commented on the unused id here, but can we remove this div entirely? (and all the other empty divs)

**Pull request merged**